File: src/prism.rb
module Kernel
  def async_require(paths, &block)
    JS::Global.window.Prism.require(block, *paths)
  end
end

module Prism
  @@instances = {}

  def self.instances
    @@instances
  end

  class Mount
    def initialize(component)
      @component = component
    end

    def render
      JSON::stringify(@component.render)
    end

    def dispatch(messageJSON)
      message = JSON::parse(messageJSON)

      instance = Prism.instances[message["instance"]]

      instance.process(message)
    end

    def event(eventJSON, id)
      DOM.event(JSON::parse(eventJSON), id)
    end
  end

  class Component
    TAG_NAMES = %i[
      a
      abbr
      address
      area
      article
      aside
      audio
      b
      base
      bdi
      bdo
      blockquote
      body
      br
      button
      canvas
      caption
      cite
      code
      col
      colgroup
      dd
      del
      details
      dfn
      dir
      div
      dl
      dt
      em
      embed
      fieldset
      figcaption
      figure
      footer
      form
      h1
      h2
      h3
      h4
      h5
      h6
      head
      header
      hgroup
      hr
      html
      i
      iframe
      img
      input
      ins
      kbd
      keygen
      label
      legend
      li
      link
      main
      map
      mark
      menu
      meta
      nav
      noscript
      object
      ol
      optgroup
      option
      p
      param
      pre
      progress
      q
      rp
      rt
      ruby
      s
      samp
      script
      section
      select
      small
      source
      span
      strong
      style
      sub
      summary
      sup
      table
      tbody
      td
      textarea
      tfoot
      th
      thead
      time
      title
      tr
      u
      ul
      video
    ]

    TAG_NAMES.each do |tag|
      define_method(tag) do |*args|
        options = {}
        className = ""
        children = []

        result = {}

        until args.empty?
          arg = args.shift

          case arg
          when String
            if arg.start_with?(".")
              className = arg
            else
              children = [text(arg)]
            end
          when Array
            children = arg
          when Object
            options = arg
          end
        end

        options.each do |key, value|
          next if value.is_a?(EventHandler) || (key.to_s).chars.first == '_'
          result[key] = value
        end

        result[:_type] = tag
        result[:_class] = className
        options[:_children] = children || []

        result[:_children] = options[:_children].compact.map do |child|
          if child.is_a?(Prism::Component)
            child.render
          elsif child.is_a?(String)
            text(child)
          else
            child
          end
        end

        result[:on] ||= {}

        options.each do |key, value|
          key_as_string = key.to_s
          next unless key_as_string.start_with?('on')
          event_name = key_as_string.sub('on', '').downcase
          result[:on][event_name] = _register_handler(value)
        end

        if options[:on]
          event_handlers = {}

          options[:on].each do |key, value|
            event_handlers[key] = value.to_hash
          end

          result[:on] = event_handlers
        end

        result
      end
    end

    def text(t)
      {:type => "text", :content => t.to_s}
    end

    def call(method_name, *args)
      EventHandler.new(self, method_name).with(*args)
    end

    def stop_propagation
      EventHandler.new(self, method_name).stop_propagation
    end

    def prevent_default
      EventHandler.new(self, method_name).prevent_default
    end

    def _register_handler(handler)
      Prism.instances[handler.id] = handler # TODO - this is a memory leak
      handler.to_hash
    end

    def render
      raise "Unimplemented render method for #{self.class.name}"
    end
  end

  class EventHandler
    attr_reader :method_name, :id

    def initialize(instance, method_name, args = [], options = {})
      @instance = instance
      @id = self.object_id
      @method_name = method_name
      @args = args
      @options = {prevent_default: false, stop_propagation: false}.merge(options)
    end

    def with(*additionalArgs)
      new_args = additionalArgs.map { |a| {type: :constant, value: a} }

      EventHandler.new(@instance, method_name, @args + new_args, @options)
    end

    def with_event
      EventHandler.new(@instance, method_name, @args + [{type: :event}], @options)
    end

    def with_event_data(*property_names)
      new_args = property_names.map { |item| {type: :event_data, key: item } }

      EventHandler.new(@instance, method_name, @args + new_args, @options)
    end

    def with_target_data(*items)
      target_args = items.map { |item| {type: :target_data, key: item } }
      EventHandler.new(@instance, method_name, @args + target_args, @options)
    end

    def prevent_default
      EventHandler.new(@instance, method_name, @args, @options.merge(prevent_default: true))
    end

    def stop_propagation
      EventHandler.new(@instance, method_name, @args, @options.merge(stop_propagation: true))
    end

    def process(message)
      call_args = []
      message_args = message["args"]

      @args.each do |arg|
        case arg[:type]
        when :constant
          call_args << arg[:value]
        else
          call_args << message_args.shift
        end
      end

      if @instance.respond_to? message["type"]
        @instance.send(message["type"], *call_args)
      else
        raise "Component #{@instance.class} has no method ##{message["type"]}"
      end
    end

    def to_hash
      {
        instance: @id,
        type: @method_name,
        args: @args,
        prevent_default: @options[:prevent_default],
        stop_propagation: @options[:stop_propagation]
      }
    end
  end
end

module DOM
  @@event_id = 0
  @@listeners = {}

  def self.get_event_id
    @@event_id += 1

    @@event_id.to_s
  end

  def self.add_listener(id, &block)
    @@listeners[id] = block
  end

  def self.select(selector)
    ElementSelection.new(selector)
  end

  def self.event(eventData, id)
    @@listeners[id].call(eventData)
  end

  class ElementSelection
    def initialize(selector)
      @selector = selector
    end

    def on(eventName, &block)
      id = DOM.get_event_id
      InternalDOM.add_event_listener(@selector, eventName, id)
      DOM.add_listener(id, &block)
    end
  end
end

module Prism
  class ExternalReferences
    @@reference_id = 0
    @@references = {}

    def self.get_ruby_reference(value)
      id = @@reference_id

      @@references[id] = value

      @@reference_id += 1

      id
    end


    def self.handle_callback(reference)
      args = []

      callable = dereference(reference)

      arg_count = InternalBindings.get_arg_count

      arg_count.times.with_index do |i|
        ref = InternalBindings.get_arg_reference(i)
        type = InternalBindings.get_type_of(ref.value)
        args << JS::Value.translate_js_value(ref, type)
      end

      callable.call(*args.slice(0...callable.parameters.size))
    end

    def self.cleanup_reference(reference)
      @@references.delete(reference)
    end

    def self.dereference(ruby_reference_id)
      @@references[ruby_reference_id]
    end

    def self.get_ruby_reference_type(prop_name, ruby_reference_id)
      value = @@references[ruby_reference_id]

      prop_name_as_int = prop_name.to_i

      if prop_name_as_int.to_s == prop_name
        prop_name = prop_name_as_int
      end

      case value[prop_name]
      when String
        "string"
      when Numeric
        "number"
      when NilClass
        "null"
      else
        fail "get_ruby_reference_type: #{prop_name} #{value[prop_name]} not handled"
      end
    end

    def self.get_ruby_reference_number(prop_name, ruby_reference_id)
      value = @@references[ruby_reference_id]

      prop_name_as_int = prop_name.to_i

      if prop_name_as_int.to_s == prop_name
        prop_name = prop_name_as_int
      end

      value[prop_name].to_f
    end

    def self.get_ruby_reference_string(prop_name, ruby_reference_id)
      value = @@references[ruby_reference_id]

      prop_name_as_int = prop_name.to_i

      if prop_name_as_int.to_s == prop_name
        prop_name = prop_name_as_int
      end

      value[prop_name].to_s
    end
  end
end

class RubyReference
  def initialize(reference:)
    @reference = reference
  end
end

class InternalBindings
  def self.args_reference
    @args_reference ||= self.get_args_reference
  end
end

module JS
  CallbackInterface = Struct.new(:name, :options)
  TypeDef = Struct.new(:name, :types)

  class Undefined
    def self.singleton
      @@instance ||= Undefined.new
    end
  end

  class Value
    def initialize(reference:)
      fail "bad reference #{reference}" unless reference.is_a?(JSReference)
      @reference = reference
    end

    def new(*args, &block)
      JS::Value.send_args_to_js(*args, &block)

      reference = InternalBindings.call_constructor(@reference.value)

      JS::Value.translate_js_value(reference, InternalBindings.get_type_of(reference.value))
    end

    def _reference
      @reference
    end

    private

    def self.translate_js_value(reference, type)
      return nil if reference.value == 0

      case type
      when "ruby_value"
        ruby_reference_id = InternalBindings.get_ruby_reference_id(reference.value)
        Prism::ExternalReferences.dereference(ruby_reference_id)
      when "object", "function"
        JS::Value.new(reference: reference)
      when "number"
        InternalBindings.get_value_number(reference.value)
      when "string"
        InternalBindings.get_value_string(reference.value)
      when "undefined"
        JS::Undefined
      when "boolean"
        value = InternalBindings.get_value_number(reference.value)

        if value.to_i == 1
          true
        else
          false
        end
      else
        fail "unhandled type #{type}"
      end
    end

    def self.send_args_to_js(*args, &block)
      InternalBindings.clear_args

      args.each_with_index do |arg, index|
        case arg
        when NilClass
          InternalBindings.set_object_null(
            InternalBindings.args_reference.value,
            index.to_s
          )
        when String
          InternalBindings.set_arg_string(index, arg)
        when JS::Value
          InternalBindings.set_arg_value(index, arg._reference.value)
        when Numeric
          InternalBindings.set_arg_number(index, arg)
        when Proc, Method
          arg_reference = Prism::ExternalReferences.get_ruby_reference(arg)

          InternalBindings.set_arg_callback(index, arg_reference)
        else
          ruby_reference_id = Prism::ExternalReferences.get_ruby_reference(arg)
          InternalBindings.set_object_ruby_value(
            InternalBindings.args_reference.value,
            index.to_s,
            ruby_reference_id
          )
        end
      end

      if block
        block_reference = Prism::ExternalReferences.get_ruby_reference(block)

        InternalBindings.set_arg_callback(args.length, block_reference)
      end
    end

    def call_function(reference, *args, &block)
      self.class.send_args_to_js(*args, &block)

      result_reference = InternalBindings.call_method_reference(
        @reference.value,
        reference.value
      )

      self.class.translate_js_value(
        result_reference,
        InternalBindings.get_type_of(result_reference.value)
      )
    end

    def call_setter(name, value)
      case value
      when String
        InternalBindings.set_object_value_string(@reference.value, name, value)
      when Numeric
        InternalBindings.set_object_value_number(@reference.value, name, value)
      when JS::Value
        InternalBindings.set_object_value(@reference.value, name, value._reference.value)
      when JS::Undefined
        InternalBindings.set_object_undefined(@reference.value, name)
      when Proc
        arg_reference = Prism::ExternalReferences.get_ruby_reference(value)

        callback_reference = InternalBindings.make_callback_reference(arg_reference)

        InternalBindings.set_object_value(@reference.value, name, callback_reference.value)
      else
        ruby_reference_id = Prism::ExternalReferences.get_ruby_reference(value)

        InternalBindings.set_object_ruby_value(@reference.value, name, ruby_reference_id)
      end
    end

    def method_missing(name, *args, &block)
      is_constructor = false

      if name.to_s.end_with?("!")
        name = name.to_s[0...-1]
        is_constructor = true
      end

      if name.to_s.end_with?("=") && args.length > 0
        value = args.first

        call_setter(name.to_s[0...-1], value)

        return value
      end

      reference = InternalBindings.get_value_reference(@reference.value, name.to_s)

      return nil if reference.value == 0

      type = InternalBindings.get_type_of(reference.value)

      if type == "function" then
        if is_constructor
          JS::Value.new(reference: reference)
        else
          call_function(reference, *args, &block)
        end
      else
        self.class.translate_js_value(reference, type)
      end
    end
  end

  class Window < Value
    def eval(*args, &block)
      method_missing(:eval, *args, &block)
    end
  end

  module Global
    def self.included(base)
      base.define_method(:window) do
        JS::Global.window
      end

      base.define_method(:document) do
        JS::Global.document
      end

      base.define_method(:undefined) do
        JS::Global.undefined
      end
    end

    def self.window
      @@window ||= JS::Window.new(reference: InternalBindings.window_reference)
    end

    def self.document
      @@document ||= JS::Value.new(reference: InternalBindings.document_reference)
    end

    def self.undefined
      JS::Undefined.singleton
    end
  end
end

CONFIG = JSON::parse(JSON_CONFIG)

File: src/prism.js
var snabbdom = require("snabbdom");
var patch = snabbdom.init([
  require("snabbdom/modules/class").default,
  require("snabbdom/modules/attributes").default,
  require("snabbdom/modules/props").default,
  require("snabbdom/modules/style").default,
  require("snabbdom/modules/eventlisteners").default,
  require("snabbdom/modules/dataset").default,
]);
var snabbdom_h = require("snabbdom/h").default;


function stringifyEvent(e) {
  const obj = {};
  for (let k in e) {
    obj[k] = e[k];
  }
  return JSON.stringify(
    obj,
    (k, v) => {
      if (v instanceof Node) return "Node";
      if (v instanceof Window) return "Window";
      return v;
    },
    " "
  );
}

function rubyVTreeToSnabbdom(rvtree) {
  if (rvtree.type === "text") {
    return rvtree.content;
  }

  let options = {};

  for (let key in rvtree) {
    if (key === "_children") {
      continue;
    }
    if (key === "_type") {
      continue;
    }
    if (key === "_class") {
      continue;
    }
    if (key === "onClick") {
      continue;
    }
    if (key === "onKeydown") {
      continue;
    }
    if (key === "onInput") {
      continue;
    }

    options[key] = rvtree[key];
  }

  if (options.on) {
    for (let key in options.on) {
      var handler = options.on[key];

      options.on[key] = function (event) {
        if (handler.prevent_default) {
          event.preventDefault();
        }

        if (handler.stop_propagation) {
          event.stopPropagation();
        }

        let args = [];

        for (let arg of handler.args) {
          if (arg.type === "event") {
            args.push(serializeEvent(event));
          }

          if (arg.type === "event_data") {
            args.push(event[arg.key]);
          }

          if (arg.type === "target_data") {
            args.push(event.target[arg.key]);
          }
        }

        var handlerWithArgs = Object.assign({}, handler, { args });

        if (handlerWithArgs.type) {
          Module.ccall(
            "dispatch",
            "void",
            ["string"],
            [JSON.stringify(handlerWithArgs)]
          );
        }
        render();
      };
    }
  }

  return snabbdom_h(
    rvtree._type + (rvtree._class || ""),
    options,
    (rvtree._children || []).map(rubyVTreeToSnabbdom)
  );
}

var currentContainer;
var allLoaded = false;
var modulesToLoad = [];

function run(element, main, config = {}) {
  currentContainer = element;
  modulesToLoad = [fetchAndLoad("/prism-ruby/prism.rb"), fetchAndLoad(main)];

  load(modulesToLoad, main, config);
}

let _refId = 1;

const references = new Map();
let args = [];

function clearArgs() {
  args.splice(0, args.length);
}

function setArgString(index, value) {
  args[index] = value;
}

function setArgNumber(index, value) {
  args[index] = value;
}

const RUBY_VALUE = Symbol("RUBY_VALUE");

let callbackArgs = [];

const registry = new FinalizationRegistry(cleanupReference);

function tryCatchThrow(f) {
  try {
    return f();
  } catch (e) {
    if (e === "longjmp") {
      throw e;
    }
    console.error(e);
    Module.ccall("print_backtrace", "void", ["string"], [e.message]);
    throw e;
  }
}

function cleanupReference(reference) {
  Module.ccall("cleanup_reference", "void", ["int"], [reference]);
}

function makeCallbackReference(callbackReference) {
  function callbackHandler(...args) {
    callbackArgs = args;
    Module.ccall("handle_callback", "void", ["int"], [callbackReference]);
  }

  callbackHandler[RUBY_VALUE] = callbackReference;

  registry.register(callbackHandler, callbackReference);

  return getReference(callbackHandler);
}

function setArgCallback(index, reference) {
  function callbackHandler(...args) {
    callbackArgs = args;
    Module.ccall("handle_callback", "void", ["int"], [reference]);
  }

  args[index] = callbackHandler;

  registry.register(callbackHandler, reference);
}

function setArgValue(index, reference) {
  const referenceValue = references.get(reference);

  tryCatchThrow(() => {
    args[index] = referenceValue;
  });
}

function setObjectValue(reference, name, value) {
  const referenceValue = references.get(reference);

  tryCatchThrow(() => {
    referenceValue[name] = value;
  });
}

function setObjectValueFromReference(reference, name, valueReference) {
  const obj = references.get(reference);
  const value = references.get(valueReference);

  tryCatchThrow(() => {
    obj[name] = value;
  });
}

function makeRubyValue(rubyReferenceId) {
  function value (...args) {
    // call my current reference
    // just return a damn ruby value, worry about primitive promotion later
    callbackArgs = args;

    const newRubyReferenceId = Module.ccall(
      "call_ruby_value_returning_reference",
      "int",
      ["int"],
      [rubyReferenceId]
    );

    const rubyType = Module.ccall(
      "get_ruby_reference_type",
      "string",
      ["int"],
      [newRubyReferenceId]
    );

    if (rubyType === 'js_value') {
      const jsReferenceId = Module.ccall(
        "get_ruby_reference_number",
        "float",
        ["string", "int"],
        ["value", rubyReferenceId]
      );

      return references.get(jsReferenceId);
    }

    return makeRubyValue(newRubyReferenceId);
  }

  value[RUBY_VALUE] = rubyReferenceId;

  const proxyMethods = {
    get: function (obj, prop) {
      if (prop === RUBY_VALUE) {
        return obj[prop];
      }

      if (prop === Symbol.iterator) {
        prop = "each";

        const rubyType = Module.ccall(
          "get_ruby_reference_property_type",
          "string",
          ["string", "int"],
          [prop, rubyReferenceId]
        );

        if (rubyType !== 'method') {
          throw new Error('Ruby value does not implement #each and cannot be iterated');
        }

        const iteratorReference = Module.ccall(
          "get_ruby_iterator_reference",
          "int",
          ["int"],
          [rubyReferenceId]
        );

        return makeRubyValue(iteratorReference);
      }

      if (Symbol.toPrimitive && prop === Symbol.toPrimitive) {
        return {}[Symbol.toPrimitive];
      }

      return tryCatchThrow(() =>{
        const rubyType = Module.ccall(
          "get_ruby_reference_property_type",
          "string",
          ["string", "int"],
          [prop, rubyReferenceId]
        );

        if (rubyType === 'number') {
          return Module.ccall(
            "get_ruby_reference_number",
            "float",
            ["string", "int"],
            [prop, rubyReferenceId]
          );
        } else if (rubyType === 'string') {
          return Module.ccall(
            "get_ruby_reference_string",
            "string",
            ["string", "int"],
            [prop, rubyReferenceId]
          );
        } else if (rubyType === 'null') {
          return null;
        } else if (rubyType === 'method') {
          const methodRubyReferenceId =  Module.ccall(
            "get_ruby_method_reference",
            "int",
            ["string", "int"],
            [prop, rubyReferenceId]
          );

          return makeRubyValue(methodRubyReferenceId);
        } else if (rubyType === 'true') {
          return true;
        } else if (rubyType === 'false') {
          return false;
        } else if (rubyType === 'js_value') {
          const jsReferenceIdString = Prism.eval(`
            Prism::ExternalReferences.get_js_value_reference_number(${JSON.stringify(prop, null, 2)}, ${rubyReferenceId})
          `);

          return references.get(parseInt(jsReferenceIdString, 10));
        } else if (rubyType === "object") {
          const newRubyReferenceId = Prism.eval(`
            Prism::ExternalReferences.get_ruby_property_object(${JSON.stringify(prop)}, ${rubyReferenceId})
          `);

          return makeRubyValue(parseInt(newRubyReferenceId, 10));
        } else {
          throw new Error('unhandled ruby type: ' + rubyType);
        }
      })
    },
  };

  const rubyValue = new Proxy(value, proxyMethods);
  registry.register(rubyValue, rubyReferenceId);
  return rubyValue;
}

function setObjectValueFromRubyReference(reference, name, rubyReferenceId) {
  const obj = references.get(reference);
  const value = makeRubyValue(rubyReferenceId);

  tryCatchThrow(() => {
    obj[name] = value;
  });
}

function getReference(obj) {
  if (obj == null) {
    return 0;
  }

  const refId = _refId++;

  references.set(refId, obj);

  return refId;
}

function freeReference(refId) {
  references.delete(refId);
}

function getWindowReference() {
  return getReference(window);
}

function getDocumentReference() {
  return getReference(document);
}

function getArgsReference() {
  return getReference(args);
}

function callMethod(reference, methodName) {
  const value = references.get(reference);

  tryCatchThrow(() => {
    if (!value) {
      throw new Error(
        `Attempted to call ${methodName} on invalid reference: ${reference}`
      );
    }

    value[methodName].apply(value, args);
  });
}

function callMethodReturningReference(thisReference, reference) {
  const thisValue = references.get(thisReference);
  const value = references.get(reference);

  return tryCatchThrow(() => {
    if (!value || !thisValue) {
      throw new Error(
        `Attempted to call invalid reference: ${thisReference}, ${reference}, ${value}`
      );
    }

    const result = value.apply(thisValue, args);

    return getReference(result);
  });
}

function callConstructor(reference) {
  const value = references.get(reference);

  return tryCatchThrow(() => {
    if (!value) {
      throw new Error(
        `Attempted to construct invalid reference: ${reference}, ${value}`
      );
    }

    const result = Reflect.construct(value, args);

    return getReference(result);
  });
}

function getValueString(reference) {
  const value = references.get(reference);

  return tryCatchThrow(() => {
    return value.toString();
  });
}

function getValueNumber(reference) {
  const value = references.get(reference);

  return tryCatchThrow(() => {
    return Number(value);
  });
}

function getValueReference(reference, property) {
  const value = references.get(reference);
  return tryCatchThrow(() => {
    // TODO - this breaks on falsey values
    if (!value) {
      throw new Error(
        `Attempted to look up ${property} on invalid reference: ${reference}`
      );
    }

    return getReference(value[property]);
  });
}

function getTypeOf(reference) {
  const value = references.get(reference);

  return tryCatchThrow(() => {
    const type = typeof value;

    if ((type === "object" || type === "function") && RUBY_VALUE in value) {
      return "ruby_value";
    }

    return typeof value;
  });
}

function getRubyReferenceId(reference) {
  const value = references.get(reference);

  return tryCatchThrow(() => {
    return value[RUBY_VALUE];
  });
}

function checkIfFunctionIsContructor(reference) {
  const value = references.get(reference);

  return tryCatchThrow(() => {
    if (value === Symbol) {
      return false;
    }

    return (
      value.prototype &&
      "constructor" in value.prototype &&
      value.prototype.constructor !== Function
    );
  });
}

function getArgCount() {
  return tryCatchThrow(() => callbackArgs.length);
}

function getArgString(index) {
  const value = callbackArgs[index];
  return tryCatchThrow(() => value.toString());
}

function getArgNumber(index) {
  return callbackArgs[index];
}

function getArgReference(index) {
  const value = callbackArgs[index];

  return tryCatchThrow(() => {
    return getReference(value);
  });
}

function getArgClassName(index) {
  const value = callbackArgs[index];

  return tryCatchThrow(() => {
    if (!value || !value.constructor) {
      return "";
    }

    return value.constructor.name;
  });
}

window.Prism = {
  run,
  eval: _eval,
  setArgString,
  setArgNumber,
  setArgCallback,
  setArgValue,
  setObjectValue,
  setObjectValueFromReference,
  setObjectValueFromRubyReference,
  freeReference,
  clearArgs,
  getValueReference,
  getValueString,
  getValueNumber,
  callMethod,
  callMethodReturningReference,
  callConstructor,
  render,
  stringifyEvent,
  getWindowReference,
  getDocumentReference,
  getArgsReference,
  getArgCount,
  getArgString,
  getArgNumber,
  getArgReference,
  getArgClassName,
  getTypeOf,
  makeCallbackReference,
  getRubyReferenceId,
  checkIfFunctionIsContructor,
  require: prismRequire,
};

function render() {
  const rvtree = JSON.parse(Module.ccall("render", "string", []));
  const vtree = rubyVTreeToSnabbdom(rvtree);

  patch(currentContainer, vtree);

  currentContainer = vtree;
}

function fetchAndLoad(name) {
  return fetch(name)
    .then((r) => r.text().then((t) => ({ ok: r.ok, text: t })))
    .then(({ ok, text }) => {
      if (!ok) {
        throw new Error(`Prism: Could not load ${name}`, text);
      }

      return { name, text };
    });
}

function _eval(s) {
  return Module.ccall("eval", "string", ["string"], [s]);
}

async function prismRequire(cb, ...paths) {
  for (const path of paths) {
    await fetchAndLoad(path).then((module) => {
      writeModule(module.name, module.text);
      Module.ccall("load_ruby", "number", ["string"], [module.name]);
    });
  }

  cb();
}

require_regex = /require(_relative)?\s*\(?\s*['|"]([^'|"]+)/;

function transformModule(moduleText) {
  const lines = moduleText.split("\n");
  const dependencies = [];

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];

    const match = line.match(require_regex);
    if (match) {
      dependencies.push(match[2]);
    } else {
      if (dependencies.length === 0) {
        return moduleText;
      }

      bodyText = lines
        .slice(i)
        .map((line) => `  ${line}`)
        .join("\n");

      const result = `async_require([${dependencies
        .map((d) => `"${d}"`)
        .join(",")}]) do\n${bodyText}\nend`;

      return result;
    }
  }

  return moduleText;
}

function writeModule(name, text) {
  const parts = name.split("/").filter((a) => a.trim() !== "");

  const directories = parts.slice(0, -1);
  const basename = parts.slice(-1)[0];

  const pwd = [];
  for (let d of directories) {
    try {
      FS.mkdir("./" + pwd.concat(d).join("/"));
    } catch (e) {}
    pwd.push(d);
  }

  FS.writeFile(`./${name}`, transformModule(text));
}

function load(modulesToLoad, main, config = {}) {
  modulePromise.then(() => {
    Promise.all(modulesToLoad).then((modules) => {
      for (let m of modules) {
        writeModule(m.name, m.text);
      }

      const result = Module.ccall(
        "load",
        "number",
        ["string", "string"],
        [main, JSON.stringify(config)]
      );
      if (result === 0) {
        //render();
      }
    });
  });
}

window.addEventListener('unhandledrejection', function (event) {
  if (event.reason === 'longjmp') {
    event.preventDefault();
    event.stopPropagation();
  }
});

const modulePromise = new Promise((resolve, reject) => {
  window.Module = {
    preRun: [],
    postRun: [resolve],
    print: (function () {
      return function (e) {
        1 < arguments.length &&
          (e = Array.prototype.slice.call(arguments).join(" ")),
          console.log(e);
      };
    })(),
    printErr: function (e) {
      1 < arguments.length &&
        (e = Array.prototype.slice.call(arguments).join(" ")),
        console.error(e);
    },
    canvas: (function () {})(),
    setStatus: function () {},
    totalDependencies: 0,
    monitorRunDependencies: function (e) {},
  };
});
